pkg/virt-controller/watch/vmi.go

```diff
@@ -1039,3 +1114,303 @@ func (c *VMIController) currentPod(vmi *virtv1.VirtualMachineInstance) (*k8sv1.P
+	if err != nil {
+		return attachmentPods, err
+	}
+
+	for _, pod := range pods {
+		ownerRef := controller.GetControllerOf(pod)
+		if ownerRef == nil || ownerRef.UID != virtlauncherPod.UID {
+			continue
+		}
+		attachmentPods = append(attachmentPods, pod)
+	}
+
+	return attachmentPods, nil
+}
+
+func (c *VMIController) needsHandleHotplug(hotplugVolumes []*virtv1.Volume, currentAttachmentPods []*k8sv1.Pod) bool {
+	// If lengths don't match, need to handle for sure. This captures single adds/deletes
+	if len(hotplugVolumes) != len(currentAttachmentPods) {
+		return true
+	}
+	volumeMap := make(map[string]*virtv1.Volume)
+	for _, volume := range hotplugVolumes {
+		volumeMap[volume.Name] = volume
+	}
+	for _, pod := range currentAttachmentPods {
+		for _, podVolume := range pod.Spec.Volumes {
+			if _, ok := volumeMap[podVolume.Name]; !ok && podVolume.VolumeSource.PersistentVolumeClaim != nil {
+				// found a pod with a PVC that is not in the hotplugged volume list, this means we need to unplug the volume
+				// This also captures the add/delete at once, no need to do extra check for the add, because if we added a
+				// volume the length check would have caught it.
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func (c *VMIController) handleHotplugVolumes(hotplugVolumes []*virtv1.Volume, hotplugAttachmentPods []*k8sv1.Pod, vmi *virtv1.VirtualMachineInstance, virtLauncherPod *k8sv1.Pod) ([]virtv1.VolumeStatus, syncError) {
+	logger := log.Log.Object(vmi)
+
+	statusCopy := vmi.Status.DeepCopy().VolumeStatus
+	// Examine pods, and determine which volumes were added, and which were deleted.
+	deletedVolumes := c.getDeletedHotplugVolumes(hotplugAttachmentPods, hotplugVolumes)
+	if len(deletedVolumes) > 0 {
+		// Some volumes were deleted, make sure we delete the hotplug pods
+		for _, volume := range deletedVolumes {
+			logger.V(1).Infof("Deleting attachment pod for volume: %s", volume.Name)
+			deletedPod, err := c.deleteAttachmentPodForVolume(vmi, volume, hotplugAttachmentPods)
+			if err != nil {
+				return nil, &syncErrorImpl{fmt.Errorf("Error deleting pod %s/%s, %v", deletedPod.Namespace, deletedPod.Name, err), FailedDeletePodReason}
+			}
+			if deletedPod != nil {
+				c.recorder.Eventf(vmi, k8sv1.EventTypeNormal, SuccessfulDeletePodReason, "Deleted hotplug attachment pod %s/%s, for volume %s", deletedPod.Namespace, deletedPod.Name, volume.Name)
+			}
+			// Not deleting volume status, updating it to indicate the pod was deleted, once the pod is really gone, the volume status will be
+			// removed.
+			for _, statusVolume := range statusCopy {
+				if statusVolume.Name == volume.Name && statusVolume.HotplugVolume != nil {
+					statusVolume.HotplugVolume.Phase = virtv1.HotplugVolumeDetaching
+					statusVolume.HotplugVolume.Reason = SuccessfulDeletePodReason
+					statusVolume.HotplugVolume.Message = fmt.Sprintf("Deleted hotplug attachment pod %s/%s, for volume %s", deletedPod.Namespace, deletedPod.Name, volume.Name)
+				}
+			}
```

Comment:
This modification of the volumeStatus is something that I'd like to see discovered in the updateStatus() function.

The pattern we're using is that status changes need to be observable based actions that the controller has performed. The vmi controller is split into two parts

- sync(), which performs actions like creating/deleting pods.
- updateStatus, which observes the state of various resources in our informers in order to calculate the status of the VMI.

These two functions are not meant to share any data between one another. Instead they work independently. The only exception to this is the `syncErr` which is used by the updateStatus function to set a condition indicating that an error occurred while syncing.